.github/workflows/check.yml

```diff
@@ -4,10 +4,24 @@ on: [push, pull_request]
 
 jobs:
   check:
-    name: run checks
+    strategy:
+      fail-fast: false
+      matrix:
+        tox-filter:
+          - "py3.8"
+          - "py3.9"
+          - "py3.10"
+          - "py3.11"
+          - "py3.12"
+          - "py3.13"
+    name: run checks ${{ matrix.tox-filter }}
     runs-on: ubuntu-latest
     steps:
       - name: checkout code
         uses: actions/checkout@v2
+      - name: build and pull images
+        run: |
+          docker compose build
+          docker compose pull --quiet
```

Comment:
this step build image for each matrix item, lets build image first and then reuse in check

Reply:
Needs to put the build on a separate job which means that the image needs pushing somewhere in between: https://docs.docker.com/build/ci/github-actions/share-image-jobs/

I've also looked at using GitHub's container registry, but that might be a challenge to use from a fork...

Reply:
I couldn't get the solution from the above link to work nicely with the `docker compose pull`. I had to either repeat the services to pull, which would cause duplication when adding and removing Postgres version, or ignore errors which could hide another real problem down the line.

Here are my attempts: https://github.com/browniebroke/django-pg-zero-downtime-migrations/actions?query=branch%3Aci%2Fseparate-build-push